### ISSUE TITLE
LBFGSB Bounded optimizer should take in arrays for bounds

### ISSUE THREAD

**Issue by hjmjohnson** [SimpleITK/SimpleITK#1733]
The LBFGSB should take in arrays for the bounds.

Providing bounds for the parameters of optimized transforms almost always requires different bounds for each parameter type.

For example, a Euler3D transform is (radians, radians, radians, mm, mm, mm)

One could easily bound to +- 15 mm for the transform, but 15 radians does not make any sense for optimization purposes.

One would like to create bounds like

```python
tendegrees= 10*pi/180

upperbound=[tendegrees, tendegrees, tendegrees, 15, 15,15]
 lowerbound= [ -x for x in upperbound] 
```

```diff
--- a/Code/Registration/src/sitkImageRegistrationMethod_CreateOptimizer.cxx
+++ b/Code/Registration/src/sitkImageRegistrationMethod_CreateOptimizer.cxx
@@ -217,28 +217,30 @@ namespace simple
       #define UPPERBOUND  2 // 10
       #define BOTHBOUND   3 // 11
 
-      unsigned char flag = NOBOUND;
       const unsigned int sitkToITK[] = {0,1,3,2};
-      if ( this->m_OptimizerLowerBound != std::numeric_limits<double>::min() )
-        {
-        flag |= LOWERBOUND;
-        }
-      if ( this->m_OptimizerUpperBound != std::numeric_limits<double>::max() )
+      _OptimizerType::BoundSelectionType boundSelection( numberOfTransformParameters );
+      for(int i =0 ; i < numberOfTransformParameters; ++i)
         {
-        flag |= UPPERBOUND;
+        unsigned char flag = NOBOUND;
+        if ( this->m_OptimizerLowerBound[i] != std::numeric_limits<double>::min() )
+          {
+          flag |= LOWERBOUND;
+          }
+        if ( this->m_OptimizerUpperBound[i] != std::numeric_limits<double>::max() )
+          {
+          flag |= UPPERBOUND;
+          }
+         boundSelection[i] = sitkToITK[flag];
         }
 
-      _OptimizerType::BoundSelectionType boundSelection( numberOfTransformParameters );
-      _OptimizerType::BoundValueType lowerBound( numberOfTransformParameters );
-      _OptimizerType::BoundValueType upperBound( numberOfTransformParameters );
-
-      boundSelection.Fill( sitkToITK[flag] );
-      lowerBound.Fill( this->m_OptimizerLowerBound );
-      upperBound.Fill( this->m_OptimizerUpperBound );
+      //_OptimizerType::BoundValueType lowerBound( numberOfTransformParameters );
+      //_OptimizerType::BoundValueType upperBound( numberOfTransformParameters );
+      //lowerBound.Fill( this->m_OptimizerLowerBound );
+      //upperBound.Fill( this->m_OptimizerUpperBound );
 
       optimizer->SetBoundSelection( boundSelection );
-      optimizer->SetLowerBound( lowerBound  );
-      optimizer->SetUpperBound( upperBound  );
+      optimizer->SetLowerBound( this->m_OptimizerLowerBound );
+      optimizer->SetUpperBound( this->m_OptimizerUpperBound  );
       optimizer->SetTrace( m_OptimizerTrace );
       optimizer->Register();

```

**Comment by blowekamp** [comment]
The motivating use case for the LBFGSB was for use with the b-spline transformation. For this application it is most useful to set the bounds for all parameters the same, hence the current interface. There are some additional considerations here with regards to memory usage, and support for multi-scale b-splines that may require some testing. Most likely we would need to overload the methods with bounds as vector parameters.

As I recall, experiments with other affine based transform and the LBFGS(B) optimizers did not yield useful results. With your suggested changes was the optimization successful?

**Comment by blowekamp** [comment]
@hjmjohnson Any further comments with respect to the above?

**Comment by hjmjohnson** [comment]
@blowekamp Sorry, I did not see your message earlier.    I am trying to implement a bounds stategey from C++ that has worked well for me in the past with a solution in simpleITK.  It requires that specific bounds be set on the optimizer ( i.e. problem domain-specific bounds).  I ran into this this roadblock, and I am now trying to implement it in python ITK version instead with conversions to/from sitk.

NOTE:  My solution is only a start.  It is incomplete.  When evaluating my time commitments I think it will be easier for me to wrap a C++ single-use program rather than fixing simpleITK to meet my needs.  I created this PR in the hopes that it will help others.

The ideal interface would be:   
```python
if isInstance(bounds, float):
    bounds = [bounds]*len(parameters)
if len(bounds) == 1:
    bounds = [ bounds[0] ]*len(parameters)
```

At a minimum, the documentation should be improved to state that the simple ITK interface does not match the ITK interface.

Thanks for your support.
